spec/deferred_spec.rb

```diff
@@ -59,4 +59,105 @@
     expect(begin_called).to eq(1)
     expect(finish_called).to eq(1)
   end
+
+  context "with delay" do
+    it "delays execution of a delayed promise" do
+      long_deferred = IOPromise::Deferred.new(0.5) { 123 }
+      expect(long_deferred).to be_pending
+
+      deferred = IOPromise::Deferred.new { 123 }
+      expect(long_deferred).to be_pending
+      expect(deferred).to be_pending
+
+      deferred.sync
+
+      expect(long_deferred).to be_pending
+      expect(deferred).to be_fulfilled
+    end
+
+    it "delays execution of concurrent delayed promises with different times" do
+      promises = []
+      promises << IOPromise::Deferred.new { Time.now }
+      promises << IOPromise::Deferred.new(0.5) { Time.now }
+      promises << IOPromise::Deferred.new(1) { Time.now }
+      promises << IOPromise::Deferred.new(1.5) { Time.now }
+      promises << IOPromise::Deferred.new(2) { Time.now }
+
+      Promise.all(promises).sync
+
+      exec_times = promises.map do |p|
+        expect(p).to be_fulfilled
+        p.value.to_f
+      end
+
+      expect(exec_times[1] - exec_times[0]).to be > 0.4
```

Comment:
Do we care about asserting the expected order of execution in this contrived example since we can?

Reply:
:+1: I've adjusted the creation order so it isn't increasing, and explicitly validated that they execute in order of the timeout.